hs/src/Reach/Eval/Core.hs

```diff
@@ -2184,13 +2184,117 @@ doBalanceInit_ fv mtok amta = do
   b <- lookupBalanceFV_ fv mtok
   doFluidSet_ b amta
 
+lookupBalances :: App FluidVar
+lookupBalances = do
+  all_toks <- readSt st_toks
+  tok_pos  <- readSt st_tok_pos
+  return $ FV_balances (fromIntegral $ length all_toks) tok_pos
+
+getBalanceOf :: Maybe DLArg -> App SLSVal
+getBalanceOf mtok = do
+  bv <- getBalanceOf' mtok
+  return $ public $ SLV_DLVar bv
+
+getBalanceOf' :: Maybe DLArg -> App DLVar
+getBalanceOf' mtok = do
+  b  <- lookupBalances
+  at <- withAt id
+  bv <- doFluidRef b
+  let env = getBalsEnv b
+  idx <- case join $ fmap (flip M.lookup env) mtok of
+          Just idx -> return (SLV_Int at $ fromIntegral idx)
+          Nothing  -> snd <$> getBalanceIndex mtok
+  tup <- doArrRef_ (snd bv) idx
+  bal <- doArrRef_ tup (SLV_Int at 1)
+  compileToVar bal
+
+getBalsEnv :: FluidVar -> M.Map DLArg Int
+getBalsEnv = \case
+  FV_balances _ m -> m
+  _ -> impossible "getBalsEnv: Expected FV_balances"
+
+networkBal :: SrcLoc -> SLVal
+networkBal at = SLV_Data at maybeTokenMap "None" $ SLV_Null at "networkBalance"
```

Comment:
Ah, I understand why you have the maybe now. I think it will be simpler to just have two paths... one for the network and one for the non-network. All these maybes are complex